tests/unit/test_probes_consumer.py

```diff
@@ -0,0 +1,137 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import json
+import unittest
+from typing import List
+
+from charms.blackbox_k8s.v0.blackbox_probes import BlackboxProbesRequirer
+from ops.charm import CharmBase
+from ops.framework import StoredState
+from ops.testing import Harness
+
+RELATION_NAME = "blackbox-probes"
+
+REQUIRER_META = f"""
+name: requirer-tester
+containers:
+  blackbox-tester:
+requires:
+  {RELATION_NAME}:
+    interface: blackbox_probes
+"""
+
+PROBES: List[dict] = [
+    {
+        "job_name": "my-first-job",
+        "params": {"module": ["http_2xx"]},
+        "static_configs": [
+            {
+                "targets": ["10.1.238.1"],
+                "labels": {"some_key": "some-value"},
+            }
+        ],
+    },
+    {
+        "job_name": "my-second-job",
+        "params": {
+            "module": ["icmp"],
+        },
+        "static_configs": [
+            {"targets": ["10.1.238.1"], "labels": {"some_other_key": "some-other-value"}}
+        ],
+    },
+]
+
+MODULES: dict = {
+    "http_2xx_longer_timeout": {
+        "prober": "http",
+        "timeout": "30s",
+    }
+}
+
+SCRAPE_METADATA = {
+    "model": "consumer-model",
+    "model_uuid": "12de4fae-06cc-4ceb-9089-567be09fec78",
+    "application": "consumer",
+    "charm_name": "test-charm",
+    "unit": "test-unit",
+}
+
+
+class BlackboxProbesRequirerCharm(CharmBase):
+    _stored = StoredState()
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args)
+        self._stored.set_default(num_events=0)
+        self.probes_consumer = BlackboxProbesRequirer(self, RELATION_NAME)
+        self.framework.observe(self.probes_consumer.on.targets_changed, self.record_events)
+
+    def record_events(self, event):
+        self._stored.num_events += 1
+
+    @property
+    def version(self):
+        return "1.0.0"
+
+
+class BlackboxProbesRequirerTest(unittest.TestCase):
+    def setUp(self):
+        self.harness = Harness(BlackboxProbesRequirerCharm, meta=REQUIRER_META)
+
+        self.addCleanup(self.harness.cleanup)
+        self.harness.begin()
+
+    def setup_charm_relations(self):
+        """Create relations used by test cases."""
+        rel_ids = []
+        self.assertEqual(self.harness.charm._stored.num_events, 0)
+        rel_id = self.harness.add_relation(RELATION_NAME, "consumer")
+        rel_ids.append(rel_id)
+        self.harness.update_relation_data(
+            rel_id,
+            "consumer",
+            {
+                "scrape_metadata": json.dumps(SCRAPE_METADATA),
+                "scrape_probes": json.dumps(PROBES),
+                "scrape_modules": json.dumps(MODULES),
+            },
+        )
+        self.assertEqual(self.harness.charm._stored.num_events, 1)
```

Comment:
This is a whitebox test. Ideally, we should test blackbox behavior.
- https://discourse.charmhub.io/t/practicing-loose-bdd-in-charm-unit-tests/6957
- https://discourse.charmhub.io/t/best-practices-for-testing-charms/15172#test-behaviors-not-methods-bdd-8